File: scripts/scrape/reclameaqui.ts
/**
 * Reclame Aqui — Playwright scraper.
 *
 * RA's `iosite` JSON endpoints are gated behind Cloudflare's "Just
 * a moment..." JS challenge. A vanilla `fetch` from a server gets
 * 403'd on every path. This script:
 *
 *   1. Launches stealth-patched Chromium
 *   2. Loads `https://www.reclameaqui.com.br/empresa/{slug}/` to
 *      pass CF's challenge once. The challenge cookie persists
 *      across the page session.
 *   3. Calls the JSON endpoint via `page.evaluate(fetch(...))` so
 *      the request inherits the page's CF cookies.
 *   4. Walks the paginated complaint list, normalizes to the
 *      Painbase Post shape, and POSTs to the ingest API.
 *
 * Run locally:
 *   pnpm tsx scripts/scrape/reclameaqui.ts nubank --dry-run
 *   pnpm tsx scripts/scrape/reclameaqui.ts nubank cashu c6bank   # batch
 *
 * Run on GHA: see .github/workflows/scrape-reclameaqui.yml.
 */
import { chromium } from 'playwright-extra'
// @ts-expect-error — no types ship with the stealth plugin
import StealthPlugin from 'puppeteer-extra-plugin-stealth'
import type { ScrapedPost } from '../../lib/types.js'
import { postScrapesToIngest } from '../../lib/ingest-client.js'

chromium.use(StealthPlugin())

const RA_BASE = 'https://www.reclameaqui.com.br'
const PAGE_SIZE = 30
const MAX_PAGES_PER_BRAND = 3 // ~90 complaints/brand/run is plenty
const NAV_DELAY_MS = 2000

interface RAComplaint {
  id?: string | number
  title?: string
  description?: string
  status?: string
  url?: string
  created?: string
  creationDate?: string
  city?: string
  state?: string
  userFirstName?: string
  userName?: string
}

interface RAComplaintsResponse {
  total?: number
  complains?: { data?: RAComplaint[] } | RAComplaint[]
  data?: RAComplaint[]
}

function normalize(c: RAComplaint, slug: string): ScrapedPost | null {
  const title = String(c.title || '').trim()
  if (!title) return null

  const url = c.url
    ? c.url.startsWith('http')
      ? c.url
      : `${RA_BASE}${c.url}`
    : c.id
      ? `${RA_BASE}/empresa/${slug}/relato/${c.id}/`
      : null
  if (!url) return null

  const created = c.creationDate || c.created
  return {
    source: 'reclameaqui',
    title: title.slice(0, 300),
    content: String(c.description || '').trim() || title,
    url,
    subreddit: slug, // Painbase reuses the subreddit slot for company name
    upvotes: 0,
    comments_count: 0,
    author: c.userFirstName || c.userName || null,
    created_at: created ? new Date(created).toISOString() : new Date().toISOString(),
  }
}

async function scrapeBrand(slug: string): Promise<ScrapedPost[]> {
  console.log(`[ra:${slug}] launching browser...`)
  const browser = await chromium.launch({
    headless: true,
    args: ['--disable-blink-features=AutomationControlled'],
  })

  // A human-ish locale + timezone makes CF's behaviour score
  // visibly happier. Brazilian locale lines up with the target audience.
  const context = await browser.newContext({
    locale: 'pt-BR',
    timezoneId: 'America/Sao_Paulo',
    viewport: { width: 1366, height: 768 },
    userAgent:
      'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36',
  })
  const page = await context.newPage()

  const out: ScrapedPost[] = []
  const seen = new Set<string>()

  try {
    // Step 1 — visit the company page to clear CF challenge.
    const companyUrl = `${RA_BASE}/empresa/${encodeURIComponent(slug)}/`
    console.log(`[ra:${slug}] resolving CF challenge via ${companyUrl}`)
    await page.goto(companyUrl, { waitUntil: 'domcontentloaded', timeout: 45_000 })
    // Heuristic wait for the CF interstitial to vanish — the stealth
    // plugin handles most cases automatically, but networks vary.
    await page.waitForTimeout(NAV_DELAY_MS)

    // If the page title is still "Just a moment..." we lost the
    // bypass. Bail and try the next brand instead of looping forever.
    const title = await page.title()
    if (/just a moment/i.test(title)) {
      console.warn(`[ra:${slug}] CF challenge not bypassed — bailing`)
      return out
    }

    // Step 2 — paginate through the JSON endpoint, inheriting the
    // CF cookie from the page context.
    for (let pageNo = 0; pageNo < MAX_PAGES_PER_BRAND; pageNo++) {
      const apiUrl = `https://iosite.reclameaqui.com.br/raichu-io-site-v1/companies/shortname/${encodeURIComponent(
        slug,
      )}/complains?count=${PAGE_SIZE}&start=${pageNo * PAGE_SIZE}&filter=created_desc`

      const json = (await page.evaluate(async (u) => {
        const r = await fetch(u, {
          credentials: 'include',
          headers: {
            accept: 'application/json, text/plain, */*',
            'accept-language': 'pt-BR,pt;q=0.9',
          },
        })
        if (!r.ok) return { __error: r.status }
        return r.json()
      }, apiUrl)) as RAComplaintsResponse | { __error: number }

      if ('__error' in json) {
        console.warn(`[ra:${slug}] page ${pageNo} → HTTP ${json.__error}, stopping`)
        break
      }

      const items: RAComplaint[] =
        (Array.isArray(json.complains) ? json.complains : json.complains?.data) ??
        json.data ??
        []
      if (!items.length) break

      let added = 0
      for (const c of items) {
        const post = normalize(c, slug)
        if (!post) continue
        const key = post.url ?? post.title
        if (seen.has(key)) continue
        seen.add(key)
        out.push(post)
        added++
      }
      console.log(`[ra:${slug}] page ${pageNo}: ${items.length} fetched, ${added} new`)

      if (items.length < PAGE_SIZE) break
      await page.waitForTimeout(1000) // throttle between pages
    }
  } finally {
    await context.close()
    await browser.close()
  }

  console.log(`[ra:${slug}] done — ${out.length} complaints`)
  return out
}

async function main() {
  // CLI args = brand slugs. Default set covers the BR fintechs we
  // care about for the v1 launch; can be overridden in the workflow
  // via inputs or by passing args to the script directly.
  const args = process.argv.slice(2).filter((a) => !a.startsWith('--'))
  const brands =
    args.length > 0
      ? args
      : ['nubank', 'cashu', 'c6-bank', 'banco-inter', 'picpay', 'mercado-pago']

  const runId = process.env.GITHUB_RUN_ID || new Date().toISOString()
  console.log(`[ra] run ${runId} — brands: ${brands.join(', ')}`)

  const all: ScrapedPost[] = []
  for (const slug of brands) {
    try {
      const posts = await scrapeBrand(slug)
      all.push(...posts)
    } catch (err) {
      console.error(`[ra:${slug}] failed:`, err instanceof Error ? err.message : err)
    }
  }

  console.log(`[ra] total scraped: ${all.length}`)

  if (all.length > 0) {
    const result = await postScrapesToIngest(all, {
      source: 'reclameaqui',
      runId,
    })
    console.log(`[ra] ingest result:`, result)
  } else {
    console.log('[ra] nothing to ingest')
  }
}

main().catch((err) => {
  console.error('[ra] fatal:', err)
  process.exit(1)
})

File: scripts/scrape/reclameaqui.py
"""Reclame Aqui scraper — Scrapling/StealthyFetcher with XHR capture.

The story so far:
    Initial attempt: hit iosite.reclameaqui.com.br directly. CF 403'd
    every request including from a browser session that had already
    cleared the challenge.

    Fix: don't fetch iosite ourselves. Visit the listing page in a
    real Camoufox browser, let the SPA make ITS OWN XHR calls to
    iosite (which carry whatever auth header the SPA generates), and
    intercept the JSON responses via Scrapling's `capture_xhr` regex.

Endpoint discovered via the probe at scripts/probe_ra_xhr.py:
    https://iosite.reclameaqui.com.br/raichu-io-site-v1/companyshowcase/
    company/{numeric_id}/items

    Returns JSON: { prev, next, data: [...], count, maxScore, aggregations }
    where each `data[]` row is a complaint with title, description,
    creationDate, status, url, etc.

Pagination: re-navigate to the listing page with ?pagina=2, ?pagina=3
each triggers a fresh XHR round we capture.

Run locally:
    python scripts/scrape/reclameaqui.py nubank --dry-run
    python scripts/scrape/reclameaqui.py nubank cashu c6-bank

Run on GHA: see .github/workflows/scrape-reclameaqui.yml.
"""
from __future__ import annotations
import json
import os
import sys
import time
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

# Ensure the package is importable when running as a script (without
# `pip install -e .`). Adds the repo root to sys.path.
sys.path.insert(0, str(Path(__file__).resolve().parents[2]))

from scrapling.fetchers import StealthyFetcher  # type: ignore[import-not-found]

from painbase_scraper import ScrapedPost, post_scrapes_to_ingest

RA_BASE = "https://www.reclameaqui.com.br"

# Number of paginated URLs to walk per brand. Each navigation = one
# capture round. RA serves ~10 complaints per page on the listing.
MAX_PAGES_PER_BRAND = 5

# XHR capture regex. Matches the JSON endpoint the SPA actually calls.
XHR_PATTERN = (
    r"iosite\.reclameaqui\.com\.br/raichu-io-site-v1/"
    r"companyshowcase/company/\d+/items"
)


def _build_complaint_url(c: dict[str, Any], slug: str) -> str | None:
    raw_url = c.get("url")
    if isinstance(raw_url, str) and raw_url:
        return raw_url if raw_url.startswith("http") else f"{RA_BASE}{raw_url}"
    cid = c.get("id")
    if cid:
        return f"{RA_BASE}/empresa/{slug}/relato/{cid}/"
    return None


def _normalize(c: dict[str, Any], slug: str) -> ScrapedPost | None:
    title = str(c.get("title") or "").strip()
    if not title:
        return None
    url = _build_complaint_url(c, slug)
    if not url:
        return None

    created_raw = (
        c.get("creationDate") or c.get("created") or c.get("lastInteractionTime")
    )
    if created_raw:
        try:
            created_iso = (
                datetime.fromisoformat(str(created_raw).replace("Z", "+00:00"))
                .astimezone(timezone.utc)
                .isoformat()
            )
        except (ValueError, TypeError):
            created_iso = datetime.now(timezone.utc).isoformat()
    else:
        created_iso = datetime.now(timezone.utc).isoformat()

    description = str(c.get("description") or "").strip() or title

    return ScrapedPost(
        source="reclameaqui",
        title=title[:300],
        content=description,
        url=url,
        # Reuse the subreddit slot for the company slug — matches the
        # convention used elsewhere so the watch dashboard's
        # "Companies" rollup works across sources.
        subreddit=slug,
        upvotes=0,
        comments_count=0,
        author=str(c.get("userFirstName") or c.get("userName") or "") or None,
        created_at=created_iso,
    )


def _extract_complaints(captured_xhr_list: list, debug_slug: str = "") -> list[dict[str, Any]]:
    """Pull complaint dicts out of every captured XHR response.

    Each `companyshowcase/company/{id}/items` response carries a
    `data` array. Concat across all captures, return raw dicts.

    When the upstream shape shifts (RA changes field names, payload
    becomes an array instead of an object, etc.) we log a 200-char
    preview of each XHR body so the next run's CI log tells us what
    to adjust without another local probe trip.
    """
    out: list[dict[str, Any]] = []
    for i, xhr in enumerate(captured_xhr_list):
        url = getattr(xhr, "url", "") or getattr(xhr, "request", {}).get("url", "") if hasattr(xhr, "request") else ""
        status = getattr(xhr, "status", "?")
        body = getattr(xhr, "body", None)
        if body is None:
            text = getattr(xhr, "text", "")
            body = text.encode("utf-8") if text else b""
        if isinstance(body, bytes):
            body_str = body.decode("utf-8", errors="replace")
        else:
            body_str = str(body)
        preview = body_str[:200].replace("\n", " ") if body_str else "(empty)"
        if not body_str:
            print(f"[ra:{debug_slug}] xhr[{i}] empty body url={url} status={status}", flush=True)
            continue
        try:
            data = json.loads(body_str)
        except json.JSONDecodeError:
            print(f"[ra:{debug_slug}] xhr[{i}] not-json status={status} preview={preview!r}", flush=True)
            continue
        if isinstance(data, list):
            # RA might have flattened the response to a bare array.
            print(f"[ra:{debug_slug}] xhr[{i}] bare-list len={len(data)} preview={preview!r}", flush=True)
            out.extend(c for c in data if isinstance(c, dict))
            continue
        if not isinstance(data, dict):
            print(f"[ra:{debug_slug}] xhr[{i}] non-dict {type(data).__name__} preview={preview!r}", flush=True)
            continue
        items = data.get("data") or data.get("items") or data.get("complaints") or data.get("results")
        if isinstance(items, list):
            print(
                f"[ra:{debug_slug}] xhr[{i}] ok keys={list(data.keys())} items={len(items)}",
                flush=True,
            )
            out.extend(c for c in items if isinstance(c, dict))
        else:
            print(
                f"[ra:{debug_slug}] xhr[{i}] dict no-list-field keys={list(data.keys())} preview={preview!r}",
                flush=True,
            )
    return out


def scrape_brand(slug: str) -> list[ScrapedPost]:
    """Pull complaints for one company slug across MAX_PAGES_PER_BRAND
    paginated visits. Returns ScrapedPost rows; never raises."""
    print(f"[ra:{slug}] launching StealthyFetcher (XHR capture)...", flush=True)

    out: list[ScrapedPost] = []
    seen: set[str] = set()

    for page_no in range(1, MAX_PAGES_PER_BRAND + 1):
        # /lista-reclamacoes/ is the SPA-rendered listing page. Each
        # navigation triggers a fresh round of iosite XHRs we capture.
        url = (
            f"{RA_BASE}/empresa/{slug}/lista-reclamacoes/"
            if page_no == 1
            else f"{RA_BASE}/empresa/{slug}/lista-reclamacoes/?pagina={page_no}"
        )

        def trigger_lazy_load(page):
            # Scroll a few viewports to make the SPA load its lazy
            # complaint cards (which is what triggers the XHR).
            for _ in range(4):
                page.mouse.wheel(0, 1500)
                page.wait_for_timeout(1000)
            page.wait_for_timeout(1500)
            return page

        try:
            resp = StealthyFetcher.fetch(
                url,
                headless=True,
                network_idle=True,
                timeout=90000,
                wait=2000,
                page_action=trigger_lazy_load,
                capture_xhr=XHR_PATTERN,
                disable_resources=True,
                solve_cloudflare=True,
                humanize=True,
            )
        except Exception as e:
            print(f"[ra:{slug}] page {page_no} fetch failed: {e}", flush=True)
            break

        if resp is None or getattr(resp, "status", None) not in (200, 304):
            print(
                f"[ra:{slug}] page {page_no} status="
                f"{getattr(resp, 'status', '?')} — stopping",
                flush=True,
            )
            break

        captured = list(getattr(resp, "captured_xhr", []) or [])
        complaints = _extract_complaints(captured, debug_slug=slug)
        if not complaints:
            print(
                f"[ra:{slug}] page {page_no} captured {len(captured)} XHRs "
                f"but 0 complaints in payload — stopping",
                flush=True,
            )
            break

        added = 0
        for c in complaints:
            post = _normalize(c, slug)
            if post is None:
                continue
            key = post.url or post.title
            if key in seen:
                continue
            seen.add(key)
            out.append(post)
            added += 1

        print(
            f"[ra:{slug}] page {page_no}: {len(captured)} XHRs, "
            f"{len(complaints)} complaints, {added} new",
            flush=True,
        )

        # If this page yielded nothing new, the SPA returned the same
        # cursor — no point continuing.
        if added == 0:
            break

        time.sleep(1.5)  # polite throttle between page navigations

    print(f"[ra:{slug}] done — {len(out)} complaints", flush=True)
    return out


def main() -> int:
    args = [a for a in sys.argv[1:] if not a.startswith("--")]
    # `cashu` and `banco-inter` returned 404 on RA — slugs apparently
    # retired or renamed. Drop until we re-probe with current ones.
    # The remaining four are confirmed live as of the last manual run.
    brands = args if args else [
        "nubank", "c6-bank", "picpay", "mercado-pago",
        "magazine-luiza", "americanas",  # BR retail giants — high complaint volume
    ]

    run_id = os.environ.get("GITHUB_RUN_ID") or datetime.now(timezone.utc).isoformat()
    print(f"[ra] run {run_id} — brands: {', '.join(brands)}", flush=True)

    all_posts: list[ScrapedPost] = []
    for slug in brands:
        try:
            all_posts.extend(scrape_brand(slug))
        except Exception as e:
            print(f"[ra:{slug}] failed: {e}", file=sys.stderr, flush=True)

    print(f"[ra] total scraped: {len(all_posts)}", flush=True)

    if not all_posts:
        print("[ra] nothing to ingest", flush=True)
        return 0

    result = post_scrapes_to_ingest(
        all_posts, source="reclameaqui", run_id=run_id,
    )
    print(
        f"[ra] ingest result: inserted={result.inserted} "
        f"duplicates={result.duplicates} skipped={result.skipped}",
        flush=True,
    )
    if result.errors:
        for err in result.errors:
            print(f"  ! {err}", flush=True)

    return 0


if __name__ == "__main__":
    sys.exit(main())
